test/BasicVaultTest.t.sol

```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import "forge-std/Test.sol";
+import "forge-std/console2.sol";
+
+import 'src/BasicVault.sol';
+import 'test/WMDTokenWithFailedTransfers.sol';
+
+abstract contract StateZero is Test {
+        
+    BasicVault public vault;
+    WMDTokenWithFailedTransfers public wmd;
+    address user;
+    address deployer;
+
+    function setUp() public virtual {
+        wmd = new WMDTokenWithFailedTransfers();
+        vault = new BasicVault(address(wmd));
+
+        user = 0x0000000000000000000000000000000000000001;
+        vm.label(user, "user");
+    }
+}
+
+contract StateZeroTest is StateZero {
+    
+    function testGetUserBalance() public {
+        console2.log("Check User has zero tokens in vault");
+        vm.prank(user);
+        uint balance = vault.balances(user);
+        assertTrue(balance == 0);
+    }
+
+    function testUserCannotWithdraw(uint amount) public {
+        console2.log("User cannot withdraw with no balance");
+        vm.assume(amount > 0);
+        vm.prank(user);
+        vm.expectRevert("Insufficient balance!");
+        vault.withdraw(amount);
+    }
+   
+    function testUserMintApproveDeposit(uint amount) public {
+        console2.log("User mints tokens and deposits into vault");
+        vm.startPrank(user);
+        vm.assume(amount > 0);
+        wmd.mint(user, amount);
+        wmd.approve(address(vault), amount);
+        vault.deposit(amount);
+        uint balance = vault.balances(user);
+        assertTrue(balance == amount);
+        vm.stopPrank();
+    }
+
+}
+
+abstract contract StateTokensMinted is StateZero {
+    uint userTokens;
+    
+    function setUp() public override virtual {
+        super.setUp();
+        
+        // state transition: user mints 100 tokens
+        userTokens = 100 * 10**18;
+        vm.prank(user);
+        wmd.mint(user, userTokens);
+    }
+
+}
+
+contract StateTokensMintedTest is StateTokensMinted {
+
+    function testUserCannotDepositWithoutApproval() public {
```

Comment:
Probably useful for you to understand how this works, but this is ERC20 functionality and you don't need to test it.